pandas/core/groupby/groupby.py

```diff
@@ -847,43 +857,70 @@ def _transform_should_cast(self, func_nm):
 
     def _cython_transform(self, how, numeric_only=True, **kwargs):
         output = collections.OrderedDict()
-        for name, obj in self._iterate_slices():
+        names = []  # type: List[Hashable]
+        for index, (name, obj) in enumerate(self._iterate_slices()):
             is_numeric = is_numeric_dtype(obj.dtype)
             if numeric_only and not is_numeric:
                 continue
 
             try:
-                result, names = self.grouper.transform(obj.values, how, **kwargs)
+                result, _ = self.grouper.transform(obj.values, how, **kwargs)
             except NotImplementedError:
                 continue
+
             if self._transform_should_cast(how):
-                output[name] = self._try_cast(result, obj)
-            else:
-                output[name] = result
+                result = self._try_cast(result, obj)
+
+            output[index] = result
+            names.append(name)
 
         if len(output) == 0:
             raise DataError("No numeric types to aggregate")
 
         return self._wrap_transformed_output(output, names)
 
-    def _wrap_aggregated_output(self, output, names=None):
+    def _wrap_aggregated_output(
+        self, output: Mapping[int, np.ndarray], names: List[Hashable]
+    ):
         raise AbstractMethodError(self)
 
-    def _wrap_transformed_output(self, output, names=None):
+    def _wrap_transformed_output(
+        self, output: Mapping[int, np.ndarray], names: List[Hashable]
+    ):
         raise AbstractMethodError(self)
 
     def _wrap_applied_output(self, keys, values, not_indexed_same=False):
         raise AbstractMethodError(self)
 
     def _cython_agg_general(self, how, alt=None, numeric_only=True, min_count=-1):
-        output = {}
+        output = collections.OrderedDict()
+        names = []  # type: List[Hashable]
+
+        # Ideally we would be able to enumerate self._iterate_slices and use
+        # the index from enumeration as the key of output, but ohlc in particular
+        # returns a (n x 4) array. Output requires 1D ndarrays as values, so we
+        # need to slice that up into 1D arrays
```

Comment:
would this be appreciably simpler if we handled ohlc separately?

Reply:
I though about this, and actually if you look back at the commits there were points in time where I did this.

Ultimately I figured for now to just keep this as generic as possible and not special case things. That does add:
  
  1. An inner loop in this method AND
  2. Complexity for `_wrap_series_output` and `_wrap_aggregated_output`, as they now *might* need to build a DataFrame instead of a Series in (I assume) < 1% of cases

The upside of punting this for now is that there are other functions that aren't entirely generic. `ohlc` converts a Series into a DataFrame, but similarly `size`, `ngroup` and `cumcount` convert a DataFrame into a Series (see the test where these have to be skipped). It might make the most sense to prevent these from going through the `generic` module entirely, but not sure yet what the best approach is